main.py

```diff
@@ -389,11 +389,20 @@ def visit_FunctionDef(self, tree):
                 T(',').join(ns.var(name) for name in arg_names),
                 T(',').join(arg_names))
         body = self.close(ns, body)
-        function_code = decoration.format(args + body)
-        return assignment_component(
-            T('{after}'),
-            T('{}, {}.__name__').format(self.store_var(tree.name), self.var(tree.name)),
-            T('{}, {!r}').format(function_code, tree.name))
+        function_code = args + body
+        if tree.decorator_list:
+            return assignment_component(
+                T('{after}'),
+                self.store_var(tree.name),
+                decoration.format(assignment_component(
+                    '__func',
+                    '__func, __func.__name__',
```

Comment:
This sets the name on the pre-decorated function; it’s up to the decorators whether to propagate the name or not.